### PR TITLE
Week4

### PR DESCRIPTION
**Pull request by kulloveth**
I'm still using 3.6.3 so not able to update new viewbinding feature to gradle file will start using it as soon as I update

### CODE REVIEW

File: Week4/app/src/main/java/com/kulloveth/moviesapp/MoviesDataManager.kt [hunk 1]
```diff
@@ -0,0 +1,225 @@
...
+/**
+ * This is a viewModel class that helps to
+ * manage data and simplify retaining data
+ * during configuration change
+ * */
+class MoviesDataManager(application: Application) : AndroidViewModel(application) {
+
+
+    private val TAG = MoviesDataManager::class.java.simpleName
+    val context = application.applicationContext
```
Comment (jenndotcodes) [inline Week4/app/src/main/java/com/kulloveth/moviesapp/MoviesDataManager.kt:26]: I like how you've incorporated architechture into this project and used the application object for the context.

File: Week4/app/src/main/java/com/kulloveth/moviesapp/MoviesDataManager.kt [hunk 2]
```diff
@@ -0,0 +1,225 @@
...
+
+    val KEY_FAVORITES = "KEY_FAVORITES"
+
+    /*
+    * keeps track of favorited movies
+    * */
+    fun isFavorite(movie: Movie): Boolean {
```
Comment (jenndotcodes) [inline Week4/app/src/main/java/com/kulloveth/moviesapp/MoviesDataManager.kt:33]: Above and beyond creating favorites, and really nice use of Kotlin here.

File: Week4/app/src/main/java/com/kulloveth/moviesapp/MoviesDataManager.kt [hunk 3]
```diff
@@ -0,0 +1,225 @@
...
+        return getFavorites(context)?.contains(movie.title) == true
+    }
+
+    // setting up movies to pass between fragments
+    private val _movieLiveData: MutableLiveData<Movie> = MutableLiveData()
```
Comment (jenndotcodes) [inline Week4/app/src/main/java/com/kulloveth/moviesapp/MoviesDataManager.kt:38]: Nice work using the LiveData

File: Week4/app/src/main/java/com/kulloveth/moviesapp/MoviesDataManager.kt [hunk 4]
```diff
@@ -0,0 +1,225 @@
...
+    private val _favoriteLiveData: MutableLiveData<List<Movie>> = MutableLiveData()
+
+
+    /*
+    * this method sorts  the movies and arrange
+    * according to its genres
+    * */
+    fun getMovieComposites(): List<CompositeItem>? {
+        val moviesByGenre = movieList.sortedBy { it.genre }
+        val genres = moviesByGenre.map { it.genre }.distinct()
```
Comment (jenndotcodes) [inline Week4/app/src/main/java/com/kulloveth/moviesapp/MoviesDataManager.kt:50]: Really nice use of the map function and distinct, good Kotlin again.

File: Week4/app/src/main/java/com/kulloveth/moviesapp/MoviesDataManager.kt [hunk 5]
```diff
@@ -0,0 +1,225 @@
...
+    fun setUpMovie(movie: Movie) {
+        _movieLiveData.value = movie
+    }
+
+    /**
+     * save list of movie [titles]
+     * with a [key] with [PreferenceManager]
+     * as a set
+     * */
+    fun saveList(key: String, titles: MutableList<String>, context: Context) {
```
Comment (jenndotcodes) [inline Week4/app/src/main/java/com/kulloveth/moviesapp/MoviesDataManager.kt:73]: Excellent job saving the movies to shared preferences.

File: Week4/app/src/main/java/com/kulloveth/moviesapp/MoviesDataManager.kt [hunk 6]
```diff
@@ -0,0 +1,225 @@
...
+            10,
+            context.getString(R.string.avenger_title),
+            context.getString(R.string.avenger_overview),
+            context.getString(R.string.avenger_release_date),
+            R.drawable.avengers,
+            context.getString(R.string.avenger_genre)
+        )
+
+    )
+}
```
Comment (jenndotcodes) [inline Week4/app/src/main/java/com/kulloveth/moviesapp/MoviesDataManager.kt:225]: Overall this class is well structured and good use of kotlin and good way of organizing the data, and nice job using the strings.xml, for this hardcoded app this is a good choice.

I'm impressed by the filter's you've used to get the lists of movies in this class.

File: Week4/app/src/main/java/com/kulloveth/moviesapp/favorites/FavoriteAdapter.kt [hunk 7]
```diff
@@ -0,0 +1,120 @@
...
+import com.kulloveth.moviesapp.R
+import com.kulloveth.moviesapp.databinding.MovieListItemsBinding
+import com.kulloveth.moviesapp.models.Movie
+
+class FavoriteAdapter(val movieItemCLickedListener: MovieItemCLickedListener) :
+    ListAdapter<Movie, FavoriteAdapter.ViewHolder>(MovieDiffCallBack()) {
+
+
+    override fun onCreateViewHolder(parent: ViewGroup, viewType: Int): ViewHolder {
+        val movieBinding =
```
Comment (jenndotcodes) [inline Week4/app/src/main/java/com/kulloveth/moviesapp/favorites/FavoriteAdapter.kt:22]: Nice job using the data binding.

File: Week4/app/src/main/java/com/kulloveth/moviesapp/favorites/FavoriteAdapter.kt [hunk 8]
```diff
@@ -0,0 +1,120 @@
...
+
+    }
+
+
+    /**
+     * A class responsible for creating and managing the view items
+     * @param movieListItemsBinding helps with easy reference to the views
+     * eliminating findViewById
+     * */
+    class ViewHolder(var movieListItemsBinding: MovieListItemsBinding) :
```
Comment (jenndotcodes) [inline Week4/app/src/main/java/com/kulloveth/moviesapp/favorites/FavoriteAdapter.kt:60]: I see you combined the viewholder class within the adapter class, this is perfectly acceptable and sometimes I prefer to do this as it makes fewer files to sift through.

File: Week4/app/src/main/java/com/kulloveth/moviesapp/favorites/FavoriteAdapter.kt [hunk 9]
```diff
@@ -0,0 +1,120 @@
...
+                }
+                val textColor =
+                    if (backgroundColor?.let { isColorDark(it) }!!) Color.WHITE else Color.BLACK
+                movieListItemsBinding.title.setTextColor(textColor)
+
+            }
+        }
+
+
+        fun isColorDark(color: Int): Boolean {
```
Comment (jenndotcodes) [inline Week4/app/src/main/java/com/kulloveth/moviesapp/favorites/FavoriteAdapter.kt:105]: This is quite a cool addition.

File: Week4/app/src/main/java/com/kulloveth/moviesapp/favorites/FavoriteFragment.kt [hunk 10]
```diff
@@ -0,0 +1,88 @@
...
+            val layoutManager =
+                GridLayoutManager(requireActivity(), 2)
+            recyclerView?.layoutManager = layoutManager
+        }
+        recyclerView?.adapter = adapter
+        moviesDataManager?.getFavoriteMovies(requireActivity())
+            ?.observe(requireActivity(), Observer {
+                Log.d("fav", "" + it)
+                adapter?.submitList(it)
+                if (it.isEmpty()) {
```
Comment (jenndotcodes) [inline Week4/app/src/main/java/com/kulloveth/moviesapp/favorites/FavoriteFragment.kt:73]: Nice job handling the case when there are no favorites.

File: Week4/app/src/main/java/com/kulloveth/moviesapp/favorites/FavoriteFragment.kt [hunk 11]
```diff
@@ -0,0 +1,88 @@
...
+                    recyclerView?.visibility = View.INVISIBLE
+                } else {
+                    binding?.noLikeTv?.visibility = View.INVISIBLE
+                    recyclerView?.visibility = View.VISIBLE
+                }
+
+            })
+    }
+
+    override fun movieItemCLicked(movie: Movie) {
```
Comment (jenndotcodes) [inline Week4/app/src/main/java/com/kulloveth/moviesapp/favorites/FavoriteFragment.kt:84]: In a future iteration you could implement this but having favorites qualifies as above and beyond already.

File: Week4/app/src/main/java/com/kulloveth/moviesapp/main/MainActivity.kt [hunk 12]
```diff
@@ -0,0 +1,27 @@
...
+        super.onCreate(savedInstanceState)
+        val binding = ActivityMainBinding.inflate(layoutInflater)
+        setContentView(binding.root)
+
+        //setup bottomNavigation with Navigation Component
+        val bottomNavView: BottomNavigationView = binding.navView;
+        val navController = Navigation.findNavController(this,
+            R.id.nav_host_fragment
+        )
+        NavigationUI.setupWithNavController(bottomNavView, navController)
```
Comment (jenndotcodes) [inline Week4/app/src/main/java/com/kulloveth/moviesapp/main/MainActivity.kt:23]: Very clever using the navigation component.

File: Week4/app/src/main/java/com/kulloveth/moviesapp/models/CompositeItem.kt [hunk 13]
```diff
@@ -0,0 +1,35 @@
+package com.kulloveth.moviesapp.models
+
+
+
+class CompositeItem {
```
Comment (jenndotcodes) [inline Week4/app/src/main/java/com/kulloveth/moviesapp/models/CompositeItem.kt:5]: This class is interesting, I hadn't thought of combining the header/movie into a compositeitem class but you've done it well in this class.

File: Week4/app/src/main/java/com/kulloveth/moviesapp/models/Movie.kt [hunk 14]
```diff
@@ -0,0 +1,21 @@
...
+@Parcelize
+data class Movie(
+    val id: Int,
+    var title: String,
+    val overview: String,
+    val releaseDate: String,
+    val image: Int,
+    val genre: String,
+    var isFavorite: Boolean = false
+) : Parcelable
```
Comment (jenndotcodes) [inline Week4/app/src/main/java/com/kulloveth/moviesapp/models/Movie.kt:21]: Nice job on this, and nice job with using Parcelable.

File: Week4/app/src/main/java/com/kulloveth/moviesapp/movies/MovieAdapter.kt [hunk 15]
```diff
@@ -0,0 +1,151 @@
...
+    }
+
+    /**
+     * this a diffutil class which uses your Object type
+     * [CompositeItem] in this class to automatically
+     * manage changes in you adapter without
+     * the need to call notifyDataSetChanged
+     * from every method updating the adapter
+     * */
+    class MovieDiffCallBack : DiffUtil.ItemCallback<CompositeItem>() {
```
Comment (jenndotcodes) [inline Week4/app/src/main/java/com/kulloveth/moviesapp/movies/MovieAdapter.kt:60]: Could these be turned into overloaded operators? This does seem to work fine though, just a theoretical question.

File: Week4/app/src/main/res/anim/item_animation.xml [hunk 16]
```diff
@@ -0,0 +1,27 @@
+<?xml version="1.0" encoding="utf-8"?>
```
Comment (jenndotcodes) [inline Week4/app/src/main/res/anim/item_animation.xml:1]: This animation really adds a nice effect.

File: Week4/app/src/main/res/drawable/genre_backgrouund.xml [hunk 17]
```diff
@@ -0,0 +1,11 @@
+<?xml version="1.0" encoding="utf-8"?>
```
Comment (jenndotcodes) [inline Week4/app/src/main/res/drawable/genre_backgrouund.xml:1]: This looks super cool in the layout.

File: Week4/app/src/main/res/drawable/ic_date_range_black_24dp.xml [hunk 18]
```diff
@@ -0,0 +1,5 @@
+<vector android:height="24dp" android:tint="@color/primaryColor"
```
Comment (jenndotcodes) [inline Week4/app/src/main/res/drawable/ic_date_range_black_24dp.xml:1]: This also looks super cool.

**Review by jenndotcodes — APPROVED** [review]
Hello! This app is fabulous. I'm awarding above and beyond. If you'd like to show it to us if we have a chance in breakout session you are invited. Really excellent job grouping the movies by genre, you've put a lot of attention to detail in this app, good commenting througout, it's really cool how you figured out if the text should be light or dark depending on the image, this looks great in landscape or portrait, nice work saving into shared preferences, excellent use of kotlin in your model classes and throughout, especially in the data manager class. Nice job using the modern view binding and navigation component. This is a fantastic app. Excellent work.

**Comment by kulloveth** [comment]
> Hello! This app is fabulous. I'm awarding above and beyond. If you'd like to show it to us if we have a chance in breakout session you are invited. Really excellent job grouping the movies by genre, you've put a lot of attention to detail in this app, good commenting througout, it's really cool how you figured out if the text should be light or dark depending on the image, this looks great in landscape or portrait, nice work saving into shared preferences, excellent use of kotlin in your model classes and throughout, especially in the data manager class. Nice job using the modern view binding and navigation component. This is a fantastic app. Excellent work.

Thank you, If there is time I will show the app

**Pull request merged**